Realm/ObjectStore/impl/async_query.hpp

```diff
@@ -0,0 +1,122 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2015 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_ASYNC_QUERY_HPP
+#define REALM_ASYNC_QUERY_HPP
+
+#include "results.hpp"
+
+#include <realm/group_shared.hpp>
+
+#include <exception>
+#include <mutex>
+#include <functional>
+#include <thread>
+#include <vector>
+
+namespace realm {
+namespace _impl {
+class AsyncQuery {
+public:
+    AsyncQuery(Results& target);
+    ~AsyncQuery();
+
+    size_t add_callback(std::function<void (std::exception_ptr)>);
+    void remove_callback(size_t token);
+
+    void unregister() noexcept;
+    void release_query() noexcept;
+
+    // Run/rerun the query if needed
+    void run();
+    // Prepare the handover object if run() did update the TableView
+    void prepare_handover();
+    // Update the target results from the handover
+    // Returns if any callbacks need to be invoked
+    bool deliver(SharedGroup& sg, std::exception_ptr err);
+    void call_callbacks();
+
+    // Attach the handed-over query to `sg`
+    void attach_to(SharedGroup& sg);
+    // Create a new query handover object and stop using the previously attached
+    // SharedGroup
+    void detatch();
+
+    Realm& get_realm() { return m_target_results->get_realm(); }
+    // Get the version of the current handover object
+    SharedGroup::VersionID version() const noexcept { return m_sg_version; }
+
+    bool is_alive() const noexcept;
+
+private:
+    // Target Results to update and a mutex which guards it
+    mutable std::mutex m_target_mutex;
```

Comment:
Which members use this mutex for synchronization? The comment mentions only `m_target_results`, but the implementation appears to also use it for `m_unregistered`.